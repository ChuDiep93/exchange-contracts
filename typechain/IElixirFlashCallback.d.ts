/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
} from "ethers";
import {
  Contract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "@ethersproject/contracts";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";

interface IElixirFlashCallbackInterface extends ethers.utils.Interface {
  functions: {
    "elixirFlashCallback(uint256,uint256,bytes)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "elixirFlashCallback",
    values: [BigNumberish, BigNumberish, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "elixirFlashCallback",
    data: BytesLike
  ): Result;

  events: {};
}

export class IElixirFlashCallback extends Contract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  on(event: EventFilter | string, listener: Listener): this;
  once(event: EventFilter | string, listener: Listener): this;
  addListener(eventName: EventFilter | string, listener: Listener): this;
  removeAllListeners(eventName: EventFilter | string): this;
  removeListener(eventName: any, listener: Listener): this;

  interface: IElixirFlashCallbackInterface;

  functions: {
    elixirFlashCallback(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "elixirFlashCallback(uint256,uint256,bytes)"(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;
  };

  elixirFlashCallback(
    fee0: BigNumberish,
    fee1: BigNumberish,
    data: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "elixirFlashCallback(uint256,uint256,bytes)"(
    fee0: BigNumberish,
    fee1: BigNumberish,
    data: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  callStatic: {
    elixirFlashCallback(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    "elixirFlashCallback(uint256,uint256,bytes)"(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    elixirFlashCallback(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "elixirFlashCallback(uint256,uint256,bytes)"(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    elixirFlashCallback(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "elixirFlashCallback(uint256,uint256,bytes)"(
      fee0: BigNumberish,
      fee1: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;
  };
}
