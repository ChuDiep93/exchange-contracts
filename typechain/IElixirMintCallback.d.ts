/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
} from "ethers";
import {
  Contract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "@ethersproject/contracts";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";

interface IElixirMintCallbackInterface extends ethers.utils.Interface {
  functions: {
    "elixirMintCallback(uint256,uint256,bytes)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "elixirMintCallback",
    values: [BigNumberish, BigNumberish, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "elixirMintCallback",
    data: BytesLike
  ): Result;

  events: {};
}

export class IElixirMintCallback extends Contract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  on(event: EventFilter | string, listener: Listener): this;
  once(event: EventFilter | string, listener: Listener): this;
  addListener(eventName: EventFilter | string, listener: Listener): this;
  removeAllListeners(eventName: EventFilter | string): this;
  removeListener(eventName: any, listener: Listener): this;

  interface: IElixirMintCallbackInterface;

  functions: {
    elixirMintCallback(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "elixirMintCallback(uint256,uint256,bytes)"(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;
  };

  elixirMintCallback(
    amount0Owed: BigNumberish,
    amount1Owed: BigNumberish,
    data: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "elixirMintCallback(uint256,uint256,bytes)"(
    amount0Owed: BigNumberish,
    amount1Owed: BigNumberish,
    data: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  callStatic: {
    elixirMintCallback(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    "elixirMintCallback(uint256,uint256,bytes)"(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    elixirMintCallback(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "elixirMintCallback(uint256,uint256,bytes)"(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    elixirMintCallback(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "elixirMintCallback(uint256,uint256,bytes)"(
      amount0Owed: BigNumberish,
      amount1Owed: BigNumberish,
      data: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;
  };
}
