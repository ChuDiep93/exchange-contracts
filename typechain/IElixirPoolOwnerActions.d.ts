/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
} from "ethers";
import {
  Contract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "@ethersproject/contracts";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";

interface IElixirPoolOwnerActionsInterface extends ethers.utils.Interface {
  functions: {
    "collectProtocol(address,uint128,uint128)": FunctionFragment;
    "initialize(address,address,uint24,int24)": FunctionFragment;
    "setFeeProtocol(uint8,uint8)": FunctionFragment;
    "setRewardRate(uint144,uint32)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "collectProtocol",
    values: [string, BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "initialize",
    values: [string, string, BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setFeeProtocol",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "setRewardRate",
    values: [BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "collectProtocol",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "initialize", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "setFeeProtocol",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRewardRate",
    data: BytesLike
  ): Result;

  events: {};
}

export class IElixirPoolOwnerActions extends Contract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  on(event: EventFilter | string, listener: Listener): this;
  once(event: EventFilter | string, listener: Listener): this;
  addListener(eventName: EventFilter | string, listener: Listener): this;
  removeAllListeners(eventName: EventFilter | string): this;
  removeListener(eventName: any, listener: Listener): this;

  interface: IElixirPoolOwnerActionsInterface;

  functions: {
    collectProtocol(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "collectProtocol(address,uint128,uint128)"(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    initialize(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "initialize(address,address,uint24,int24)"(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    setFeeProtocol(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "setFeeProtocol(uint8,uint8)"(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    setRewardRate(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "setRewardRate(uint144,uint32)"(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<ContractTransaction>;
  };

  collectProtocol(
    recipient: string,
    amount0Requested: BigNumberish,
    amount1Requested: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "collectProtocol(address,uint128,uint128)"(
    recipient: string,
    amount0Requested: BigNumberish,
    amount1Requested: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  initialize(
    _token0: string,
    _token1: string,
    _fee: BigNumberish,
    _tickSpacing: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "initialize(address,address,uint24,int24)"(
    _token0: string,
    _token1: string,
    _fee: BigNumberish,
    _tickSpacing: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  setFeeProtocol(
    feeProtocol0: BigNumberish,
    feeProtocol1: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "setFeeProtocol(uint8,uint8)"(
    feeProtocol0: BigNumberish,
    feeProtocol1: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  setRewardRate(
    rewardPerSecondX48: BigNumberish,
    rewardRateEffectiveUntil: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "setRewardRate(uint144,uint32)"(
    rewardPerSecondX48: BigNumberish,
    rewardRateEffectiveUntil: BigNumberish,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  callStatic: {
    collectProtocol(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: CallOverrides
    ): Promise<{
      amount0: BigNumber;
      amount1: BigNumber;
      0: BigNumber;
      1: BigNumber;
    }>;

    "collectProtocol(address,uint128,uint128)"(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: CallOverrides
    ): Promise<{
      amount0: BigNumber;
      amount1: BigNumber;
      0: BigNumber;
      1: BigNumber;
    }>;

    initialize(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    "initialize(address,address,uint24,int24)"(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    setFeeProtocol(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    "setFeeProtocol(uint8,uint8)"(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    setRewardRate(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    "setRewardRate(uint144,uint32)"(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    collectProtocol(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "collectProtocol(address,uint128,uint128)"(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    initialize(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "initialize(address,address,uint24,int24)"(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    setFeeProtocol(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "setFeeProtocol(uint8,uint8)"(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    setRewardRate(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "setRewardRate(uint144,uint32)"(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    collectProtocol(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "collectProtocol(address,uint128,uint128)"(
      recipient: string,
      amount0Requested: BigNumberish,
      amount1Requested: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    initialize(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "initialize(address,address,uint24,int24)"(
      _token0: string,
      _token1: string,
      _fee: BigNumberish,
      _tickSpacing: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    setFeeProtocol(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "setFeeProtocol(uint8,uint8)"(
      feeProtocol0: BigNumberish,
      feeProtocol1: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    setRewardRate(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "setRewardRate(uint144,uint32)"(
      rewardPerSecondX48: BigNumberish,
      rewardRateEffectiveUntil: BigNumberish,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;
  };
}
