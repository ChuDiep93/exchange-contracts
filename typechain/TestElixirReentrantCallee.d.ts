/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import {
  ethers,
  EventFilter,
  Signer,
  BigNumber,
  BigNumberish,
  PopulatedTransaction,
} from "ethers";
import {
  Contract,
  ContractTransaction,
  Overrides,
  CallOverrides,
} from "@ethersproject/contracts";
import { BytesLike } from "@ethersproject/bytes";
import { Listener, Provider } from "@ethersproject/providers";
import { FunctionFragment, EventFragment, Result } from "@ethersproject/abi";

interface TestElixirReentrantCalleeInterface extends ethers.utils.Interface {
  functions: {
    "elixirSwapCallback(int256,int256,bytes)": FunctionFragment;
    "swapToReenter(address)": FunctionFragment;
  };

  encodeFunctionData(
    functionFragment: "elixirSwapCallback",
    values: [BigNumberish, BigNumberish, BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "swapToReenter",
    values: [string]
  ): string;

  decodeFunctionResult(
    functionFragment: "elixirSwapCallback",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "swapToReenter",
    data: BytesLike
  ): Result;

  events: {};
}

export class TestElixirReentrantCallee extends Contract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  on(event: EventFilter | string, listener: Listener): this;
  once(event: EventFilter | string, listener: Listener): this;
  addListener(eventName: EventFilter | string, listener: Listener): this;
  removeAllListeners(eventName: EventFilter | string): this;
  removeListener(eventName: any, listener: Listener): this;

  interface: TestElixirReentrantCalleeInterface;

  functions: {
    elixirSwapCallback(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "elixirSwapCallback(int256,int256,bytes)"(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    swapToReenter(
      pool: string,
      overrides?: Overrides
    ): Promise<ContractTransaction>;

    "swapToReenter(address)"(
      pool: string,
      overrides?: Overrides
    ): Promise<ContractTransaction>;
  };

  elixirSwapCallback(
    arg0: BigNumberish,
    arg1: BigNumberish,
    arg2: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "elixirSwapCallback(int256,int256,bytes)"(
    arg0: BigNumberish,
    arg1: BigNumberish,
    arg2: BytesLike,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  swapToReenter(
    pool: string,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  "swapToReenter(address)"(
    pool: string,
    overrides?: Overrides
  ): Promise<ContractTransaction>;

  callStatic: {
    elixirSwapCallback(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    "elixirSwapCallback(int256,int256,bytes)"(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: CallOverrides
    ): Promise<void>;

    swapToReenter(pool: string, overrides?: CallOverrides): Promise<void>;

    "swapToReenter(address)"(
      pool: string,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {};

  estimateGas: {
    elixirSwapCallback(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;

    "elixirSwapCallback(int256,int256,bytes)"(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<BigNumber>;

    swapToReenter(pool: string, overrides?: Overrides): Promise<BigNumber>;

    "swapToReenter(address)"(
      pool: string,
      overrides?: Overrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    elixirSwapCallback(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "elixirSwapCallback(int256,int256,bytes)"(
      arg0: BigNumberish,
      arg1: BigNumberish,
      arg2: BytesLike,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    swapToReenter(
      pool: string,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;

    "swapToReenter(address)"(
      pool: string,
      overrides?: Overrides
    ): Promise<PopulatedTransaction>;
  };
}
