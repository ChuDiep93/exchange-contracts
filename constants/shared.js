exports.FOUNDATION_MULTISIG = {
  owners: [
  "0x0000B0907d933104A582B399128502f62a48F3ac",
  "0x72C397908Cb93d1B569BBB0Ff8d3D26B7b21d730",
  "0xFCE4a07f5786c255B36D463DAeFBaA3CB42fbf03",
  "0xd1Bc4131BfE891A42C2756e6a80d052e25bbefeC",
  "0x372E6c7CD93Df13A799A390b7DDa7C211abA63B4"
  ],
  threshold: 3
};
