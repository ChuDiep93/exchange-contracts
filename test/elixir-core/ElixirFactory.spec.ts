import { Wallet } from "ethers";
import { ethers, waffle, network } from "hardhat";
import { ElixirFactory } from "../../typechain/ElixirFactory";
import { expect } from "./shared/expect";
import snapshotGasCost from "./shared/snapshotGasCost";
import "dotenv/config";

import {
  FeeAmount,
  getCreate2Address,
  getCreate2AddressWithInitCodeHash,
  TICK_SPACINGS,
} from "./shared/utilities";

const { constants } = ethers;

const TEST_ADDRESSES: [string, string] = [
  "0x1000000000000000000000000000000000000000",
  "0x2000000000000000000000000000000000000000",
];

const POOL_IMPLEMENTATION = "0x5cB5539A18591947C82f5D840B05ed79f6395491";
const POOL_INIT_CODE_HASH =
  "0x41a723f9e6457830b1b7a44df4435fab88581d073226894b33131815dd674c22";

const createFixtureLoader = waffle.createFixtureLoader;

describe("ElixirFactory", () => {
  let wallet: Wallet, other: Wallet;

  let factory: ElixirFactory;
  let poolBytecode: string;
  const fixture = async () => {
    //  impersonating poolDeployer's account
    const poolDeployerAddress = "0x427207B1Cdb6F2Ab8B1D21Ab77600f00b0a639a7";
    await network.provider.request({
      method: "hardhat_impersonateAccount",
      params: [poolDeployerAddress],
    });

    const poolDeployer = await (ethers as any).getSigner(poolDeployerAddress);

    //  fund the impersonated account
    await wallet.sendTransaction({
      to: poolDeployerAddress,
      value: ethers.utils.parseEther("100"),
    });

    const poolFactory = await ethers.getContractFactory("ElixirPool");
    const poolImplementation = await poolFactory.connect(poolDeployer).deploy();

    const factoryFactory = await ethers.getContractFactory("ElixirFactory");
    const factory = await factoryFactory.deploy(POOL_IMPLEMENTATION);
    return factory as ElixirFactory;
  };

  let loadFixture: ReturnType<typeof createFixtureLoader>;
  before("create fixture loader", async () => {
    [wallet, other] = await (ethers as any).getSigners();

    loadFixture = createFixtureLoader([wallet, other]);
  });

  before("load pool bytecode", async () => {
    poolBytecode = (await ethers.getContractFactory("ElixirPool")).bytecode;
  });

  beforeEach("deploy factory", async () => {
    factory = await loadFixture(fixture);
  });

  it("implementation address", async () => {
    expect(await factory.implementation()).to.be.eq(POOL_IMPLEMENTATION);
  });

  it("owner is deployer", async () => {
    expect(await factory.owner()).to.eq(wallet.address);
  });

  it("factory bytecode size", async () => {
    expect(
      ((await waffle.provider.getCode(factory.address)).length - 2) / 2
    ).to.matchSnapshot();
  });

  it("pool bytecode size", async () => {
    await factory.createPool(
      TEST_ADDRESSES[0],
      TEST_ADDRESSES[1],
      FeeAmount.MEDIUM
    );
    const poolAddress = getCreate2Address(
      factory.address,
      TEST_ADDRESSES,
      FeeAmount.MEDIUM,
      poolBytecode
    );
    expect(
      ((await waffle.provider.getCode(poolAddress)).length - 2) / 2
    ).to.matchSnapshot();
  });

  it("initial enabled fee amounts", async () => {
    expect(await factory.feeAmountTickSpacing(FeeAmount.LOW)).to.eq(
      TICK_SPACINGS[FeeAmount.LOW]
    );
    expect(await factory.feeAmountTickSpacing(FeeAmount.MEDIUM)).to.eq(
      TICK_SPACINGS[FeeAmount.MEDIUM]
    );
    expect(await factory.feeAmountTickSpacing(FeeAmount.HIGH)).to.eq(
      TICK_SPACINGS[FeeAmount.HIGH]
    );
  });

  async function createAndCheckPool(
    tokens: [string, string],
    feeAmount: FeeAmount,
    tickSpacing: number = TICK_SPACINGS[feeAmount]
  ) {
    const create2Address = getCreate2AddressWithInitCodeHash(
      factory.address,
      tokens,
      feeAmount,
      POOL_INIT_CODE_HASH
    );
    const create = factory.createPool(tokens[0], tokens[1], feeAmount);

    await expect(create)
      .to.emit(factory, "PoolCreated")
      .withArgs(
        TEST_ADDRESSES[0],
        TEST_ADDRESSES[1],
        feeAmount,
        tickSpacing,
        create2Address
      );

    await expect(factory.createPool(tokens[0], tokens[1], feeAmount)).to.be
      .reverted;
    await expect(factory.createPool(tokens[1], tokens[0], feeAmount)).to.be
      .reverted;
    expect(
      await factory.getPool(tokens[0], tokens[1], feeAmount),
      "getPool in order"
    ).to.eq(create2Address);
    expect(
      await factory.getPool(tokens[1], tokens[0], feeAmount),
      "getPool in reverse"
    ).to.eq(create2Address);

    const poolContractFactory = await ethers.getContractFactory("ElixirPool");
    const pool = poolContractFactory.attach(create2Address);

    expect(await pool.factory(), "pool factory address").to.eq(factory.address);
    expect(await pool.token0(), "pool token0").to.eq(TEST_ADDRESSES[0]);
    expect(await pool.token1(), "pool token1").to.eq(TEST_ADDRESSES[1]);
    expect(await pool.fee(), "pool fee").to.eq(feeAmount);
    expect(await pool.tickSpacing(), "pool tick spacing").to.eq(tickSpacing);
  }

  describe("#createPool", () => {
    it("succeeds for low fee pool", async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.LOW);
    });

    it("succeeds for medium fee pool", async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.MEDIUM);
    });
    it("succeeds for high fee pool", async () => {
      await createAndCheckPool(TEST_ADDRESSES, FeeAmount.HIGH);
    });

    it("succeeds if tokens are passed in reverse", async () => {
      await createAndCheckPool(
        [TEST_ADDRESSES[1], TEST_ADDRESSES[0]],
        FeeAmount.MEDIUM
      );
    });

    it("fails if token a == token b", async () => {
      await expect(
        factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[0], FeeAmount.LOW)
      ).to.be.reverted;
    });

    it("fails if token a is 0 or token b is 0", async () => {
      await expect(
        factory.createPool(
          TEST_ADDRESSES[0],
          constants.AddressZero,
          FeeAmount.LOW
        )
      ).to.be.reverted;
      await expect(
        factory.createPool(
          constants.AddressZero,
          TEST_ADDRESSES[0],
          FeeAmount.LOW
        )
      ).to.be.reverted;
      await expect(
        factory.createPool(
          constants.AddressZero,
          constants.AddressZero,
          FeeAmount.LOW
        )
      ).to.be.revertedWith("");
    });

    it("fails if fee amount is not enabled", async () => {
      await expect(
        factory.createPool(TEST_ADDRESSES[0], TEST_ADDRESSES[1], 250)
      ).to.be.reverted;
    });

    it("gas", async () => {
      await snapshotGasCost(
        factory.createPool(
          TEST_ADDRESSES[0],
          TEST_ADDRESSES[1],
          FeeAmount.MEDIUM
        )
      );
    });
  });

  describe("#setOwner", () => {
    it("fails if caller is not owner", async () => {
      await expect(factory.connect(other).setOwner(wallet.address)).to.be
        .reverted;
    });

    it("updates owner", async () => {
      await factory.setOwner(other.address);
      expect(await factory.owner()).to.eq(other.address);
    });

    it("emits event", async () => {
      await expect(factory.setOwner(other.address))
        .to.emit(factory, "OwnerChanged")
        .withArgs(wallet.address, other.address);
    });

    it("cannot be called by original owner", async () => {
      await factory.setOwner(other.address);
      await expect(factory.setOwner(wallet.address)).to.be.reverted;
    });
  });

  describe("#enableFeeAmount", () => {
    it("fails if caller is not owner", async () => {
      await expect(factory.connect(other).enableFeeAmount(100, 2)).to.be
        .reverted;
    });
    it("fails if fee is too great", async () => {
      await expect(factory.enableFeeAmount(1000000, 10)).to.be.reverted;
    });
    it("fails if tick spacing is too small", async () => {
      await expect(factory.enableFeeAmount(500, 0)).to.be.reverted;
    });
    it("fails if tick spacing is too large", async () => {
      await expect(factory.enableFeeAmount(500, 16834)).to.be.reverted;
    });
    it("fails if already initialized", async () => {
      await factory.enableFeeAmount(100, 5);
      await expect(factory.enableFeeAmount(100, 10)).to.be.reverted;
    });
    it("sets the fee amount in the mapping", async () => {
      await factory.enableFeeAmount(100, 5);
      expect(await factory.feeAmountTickSpacing(100)).to.eq(5);
    });
    it("emits an event", async () => {
      await expect(factory.enableFeeAmount(100, 5))
        .to.emit(factory, "FeeAmountEnabled")
        .withArgs(100, 5);
    });
    it("enables pool creation", async () => {
      await factory.enableFeeAmount(250, 15);
      await createAndCheckPool([TEST_ADDRESSES[0], TEST_ADDRESSES[1]], 250, 15);
    });
  });
});
