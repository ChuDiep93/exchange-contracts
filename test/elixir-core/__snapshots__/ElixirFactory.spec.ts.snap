// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElixirFactory #createPool gas 1`] = `205955`;

exports[`ElixirFactory factory bytecode size 1`] = `1766`;

exports[`ElixirFactory pool bytecode size 1`] = `0`;
